crates/ide_completion/src/context.rs

```diff
@@ -290,23 +292,114 @@ impl<'a> CompletionContext<'a> {
         file_with_fake_ident: SyntaxNode,
         offset: TextSize,
     ) {
-        // FIXME: this is wrong in at least two cases:
-        //  * when there's no token `foo($0)`
-        //  * when there is a token, but it happens to have type of it's own
-        self.expected_type = self
-            .token
-            .ancestors()
-            .find_map(|node| {
-                let ty = match_ast! {
+        let expected = {
+            let mut node = self.token.parent();
+            loop {
+                let ret = match_ast! {
                     match node {
-                        ast::Pat(it) => self.sema.type_of_pat(&it),
-                        ast::Expr(it) => self.sema.type_of_expr(&it),
-                        _ => return None,
+                        ast::LetStmt(it) => {
+                            cov_mark::hit!(expected_type_let_with_leading_char);
+                            cov_mark::hit!(expected_type_let_without_leading_char);
+                            let ty = it.pat()
+                                .map(|pat| self.sema.type_of_pat(&pat))
+                                .flatten();
```

Comment:
Is this

```suggestion
                                .and_then(|pat| self.sema.type_of_pat(&pat));
```

perhaps?

Reply:
:+1: I've updated this in all places.